File: LeetCode/2075.Decode-the-Slanted-Ciphertext.cpp
// 2075. Decode the Slanted Ciphertext

class Solution {
public:
    string decodeCiphertext(string encodedText, int rows) {
        if(rows == 1 || encodedText.length()==0) return encodedText;
        int n = encodedText.size();
        int cols = encodedText.length()/rows;

        string ans = "";
        for(int i=0; i<cols; i++) {
            int r=0, c=i;
            while(r < rows && c < cols) {
                ans += encodedText[r*cols + c];
                r++;
                c++;
            }
        }

        while(ans.length() > 0 && ans.back() == ' ') ans.pop_back();

        return ans;
    }
};

File: LeetCode/2075-Decode-the-Slanted-Ciphertext.java
// 2075. Decode the Slanted Ciphertext

class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        int n = encodedText.length();
        int cols = n / rows;
    
        char[][] grid = new char[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                grid[i][j] = encodedText.charAt((i * cols) + j);
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int j = 0; j < cols; j++) {
            for (int i = 0; i < rows; i++) {
                if(i + j < cols) sb.append(grid[i][i + j]);
            }
        }
        while (!sb.isEmpty() && sb.charAt(sb.length() - 1) == ' ') sb.deleteCharAt(sb.length() - 1);

        return sb.toString();
    }
}
